### ISSUE TITLE
Update project preferences

### ISSUE THREAD

**Issue by goanpeca** [spyder-ide/spyder#9804]
- [ ] Include support for configuration setup (Global vs. Project)
- [ ] Hide configuration that do not make sense to display for projects (if user selects Project level config)
- [ ] Include project configuration on the preferences

**Comment by ccordoba12** [comment]
@dalthviz, did you find the notes I gave you in January for this?

**Comment by dalthviz** [comment]
@ccordoba12 I searched but I didn't find them :/

**Comment by ccordoba12** [comment]
Ok, no worries. I'll try to remember what I was thinking about it.

**Comment by goanpeca** [comment]
@ccordoba12 could you paste the paper here :-) ? thanks!

**Comment by goanpeca** [comment]
I still need to update icons and the actual logic for environments (a separate PR for anaconda-project). Is this what you had in mind @ccordoba12?


<img width="861" alt="Screen Shot 2019-07-22 at 12 29 49" src="https://user-images.githubusercontent.com/3627835/61669635-5b6e1380-aca6-11e9-91d4-0807f313c8f6.png">

<img width="857" alt="Screen Shot 2019-07-22 at 12 30 12" src="https://user-images.githubusercontent.com/3627835/61669633-5b6e1380-aca6-11e9-8ecf-d18d92292435.png">

<img width="860" alt="Screen Shot 2019-07-22 at 12 30 03" src="https://user-images.githubusercontent.com/3627835/61669634-5b6e1380-aca6-11e9-9572-3ce9c1e35a4d.png">

<img width="858" alt="Screen Shot 2019-07-22 at 12 30 17" src="https://user-images.githubusercontent.com/3627835/61669632-5b6e1380-aca6-11e9-982d-425f723e7539.png">

**Comment by goanpeca** [comment]
We could also add something related to `Version Control` which could be a separate page, or be part of the general one?

**Comment by goanpeca** [comment]
@jnsebgosselin do you have some suggestions for the icons we could use :-) ?

**Comment by ccordoba12** [comment]
> Is this what you had in mind

Yes, this is pretty much what I had in mind! Thanks!

About your screenshots:

* The "Variables" entry should be renamed to "Environment variables" to be quite clear about the meaning of that.
* The "Environment and packages" entry should be renamed simply to "Conda enviroment" and it should be deactivated (or better still, not shown at all) if Spyder is not running under Anaconda. Besides, that entry should have two tabs: one called "Environment" and the other one called "Packages". I think in "Environment" we should only show "Use existing conda environment" (without the `base` env on that list) and "Project environment". The idea with that is to force people to not use the base env, which is a killer for reproducibility. And the "Packages" tab should show the table of packages you already created.
* Perhaps "Master file" should be removed as an entry and moved as a tab in "General"? What do you think?

About your questions:

* > We could also add something related to Version Control which could be a separate page, or be part of the general one?

    I think we should add a separate page (entry) for this.

* > Where should we include the Use Global Preferences vs Use Project Preferences option? in the general page?

Yes, I think we should add this to "General".

**Comment by goanpeca** [comment]
>The "Variables" entry should be renamed to "Environment variables" to be quite clear about the meaning of that.

👍 

>The "Environment and packages" entry should be renamed simply to "Conda enviroment" and it should be deactivated (or better still, not shown at all) if Spyder is not running under Anaconda. Besides, that entry should have two tabs: one called "Environment" and the other one called "Packages". I think in "Environment" we should only show "Use existing conda environment" (without the base env on that list) and "Project environment". The idea with that is to force people to not use the base env, which is a killer for reproducibility. And the "Packages" tab should show the table of packages you already created.

Ok, but I need some extra pointers on how we will handle some things. What packages should we display on the packages tab when:

- Project env: This is clear that we will track only the packages the user explicitly asked for.
- Use existing conda environment: What packages here? All the packages that currently are installed on the env? or?


>Perhaps "Master file" should be removed as an entry and moved as a tab in "General"? What do you think?

Yes I think it makes sense for now.

>We could also add something related to Version Control which could be a separate page, or be part of the general one? I think we should add a separate page (entry) for this.

Ok will do.

**Comment by ccordoba12** [comment]
> Use existing conda environment: What packages here? All the packages that currently are installed on the env? or?

I agree this is the crucial point. Without users letting us know a specific set of packages in their environment, it's not possible for us to guarantee project reproducibility.

I see two options here:

1. We should discourage the usage of previously created conda envs at project creation and also when users select "Use existing conda environment". We could do that by showing a modal dialog, saying that it's not possible to guarantee reproducibility of the project in that case.
2. Disabling the "Packages" tab, or showing the entire list of packages in the env along with a warning instructing users to select the main dependencies of the project (e.g. Numpy, Pandas, Matplotlib, and not libgcc-ng) so we can register them in `anaconda-project.yml`.

**Comment by ccordoba12** [comment]
Perhaps both options apply?

**Comment by goanpeca** [comment]
>Disabling the "Packages" tab, or showing the entire list of packages in the env along with a warning instructing users to select the main dependencies of the project (e.g. Numpy, Pandas, Matplotlib, and not libgcc-ng) so we can register them in anaconda-project.yml.

The issue I see with this option (besides adding greater complexity) is that two or more projects could be using this other conda environment so, if the user has different dependencies on different projects, that conda env might reach a point where it will has to update some packages on project switch (because the user selected different "packages" as the main packages of their environment). 

The only "solution" to this is that a project could only use another existing conda env, but this also probably defeat its purpose for users wanting to recycle the same env for may different projects.

>Disabling the "Packages" tab, or showing the entire list of packages

What I think would make sense to me is that if the user select an existing environment, we can display the whole list (as information for the user) but disable any modification of that env from the spyder interface (remove or disable the add/edit/remove). This allows power users that want to recycle envs to use them, but enforces "best practices" for non power users.

Thoughts?

**Comment by ccordoba12** [comment]
> The issue I see with this option (besides adding greater complexity) is that two or more projects could be using this other conda environment so, if the user has different dependencies on different projects, that conda env might reach a point where it will has to update some packages on project switch (because the user selected different "packages" as the main packages of their environment).

Great point! I haven't thought about that!

> What I think would make sense to me is that if the user select an existing environment, we can display the whole list (as information for the user) but disable any modification of that env from the spyder interface (remove or disable the add/edit/remove). This allows power users that want to recycle envs to use them, but enforces "best practices" for non power users.

Agreed, this is a very good option! Besides, we should show a QLabel with a warning saying that the project is not reproducible.

**Comment by goanpeca** [comment]
![pref](https://user-images.githubusercontent.com/3627835/61731399-016e5c00-ad41-11e9-8f92-e74a2f9e38da.gif)

**Comment by ccordoba12** [comment]
Looking good, thanks! My only recommendation (for now) is to remove the QGroupBoxes called "Environment packages" and "Conda environment" because now those are unnecessary with the new tabs.

**Comment by mrclary** [comment]
@ccordoba12 @goanpeca, it seems to me that there are elements here that we can extract to the Environment Manager plugin #17654.

Also, is there any consideration for having multiple projects open at one time? But maybe this is not necessary if IPython consoles are not restarted when switching projects.

**Comment by ccordoba12** [comment]
> it seems to me that there are elements here that we can extract to the Environment Manager

Agreed. We need to discuss further about it.
